File: Sources/JetStreamWireFormat/Collections.swift
// JetStream WireFormat — Collection Types
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

import Foundation

// MARK: - Array (Vec)

// r[impl jetstream.wireformat.vec]
// r[impl jetstream.wireformat.swift.array]
extension Array: WireFormat where Element: WireFormat {
    public func byteSize() -> UInt32 {
        var size: UInt32 = 2 // u16 count prefix
        for element in self {
            size += element.byteSize()
        }
        return size
    }

    public func encode(writer: inout BinaryWriter) throws {
        guard self.count <= Int(UInt16.max) else {
            throw WireFormatError.tooManyElements(self.count)
        }
        try UInt16(self.count).encode(writer: &writer)
        for element in self {
            try element.encode(writer: &writer)
        }
    }

    public static func decode(reader: inout BinaryReader) throws -> [Element] {
        let count = try UInt16.decode(reader: &reader)
        var result: [Element] = []
        result.reserveCapacity(Int(count))
        for _ in 0..<count {
            result.append(try Element.decode(reader: &reader))
        }
        return result
    }
}

// MARK: - WireData (Data with u32 length prefix)

// r[impl jetstream.wireformat.data]
// r[impl jetstream.wireformat.swift.data]
/// A byte buffer encoded with a u32 length prefix (unlike `[UInt8]` which uses u16).
/// Maximum length is 32 MB (33,554,432 bytes).
public struct WireData: Equatable {
    public var bytes: Data

    public init(_ bytes: Data = Data()) {
        self.bytes = bytes
    }

    public init(_ bytes: [UInt8]) {
        self.bytes = Data(bytes)
    }
}

private let maxDataLength: UInt32 = 32 * 1024 * 1024

extension WireData: WireFormat {
    public func byteSize() -> UInt32 {
        4 + UInt32(bytes.count)
    }

    public func encode(writer: inout BinaryWriter) throws {
        try UInt32(bytes.count).encode(writer: &writer)
        writer.writeBytes(bytes)
    }

    public static func decode(reader: inout BinaryReader) throws -> WireData {
        let length = try UInt32.decode(reader: &reader)
        guard length <= maxDataLength else {
            throw WireFormatError.dataTooLarge(length)
        }
        guard reader.remaining >= Int(length) else {
            throw WireFormatError.unexpectedEOF
        }
        let bytes = try reader.readBytes(count: Int(length))
        return WireData(bytes)
    }
}

// MARK: - OrderedMap (BTreeMap equivalent)

// r[impl jetstream.wireformat.map]
// r[impl jetstream.wireformat.swift.dict]
/// An ordered map that encodes key-value pairs sorted by key.
/// Uses a sorted array of tuples internally.
public struct OrderedMap<K: WireFormat & Comparable, V: WireFormat>: Equatable
    where K: Equatable, V: Equatable
{
    public var entries: [(key: K, value: V)]

    public init() {
        self.entries = []
    }

    public init(_ entries: [(key: K, value: V)]) {
        self.entries = entries.sorted { $0.key < $1.key }
    }

    public static func == (lhs: OrderedMap, rhs: OrderedMap) -> Bool {
        guard lhs.entries.count == rhs.entries.count else { return false }
        for (l, r) in zip(lhs.entries, rhs.entries) {
            if l.key != r.key || l.value != r.value { return false }
        }
        return true
    }
}

extension OrderedMap: WireFormat {
    public func byteSize() -> UInt32 {
        var size: UInt32 = 2
        for (key, value) in entries {
            size += key.byteSize() + value.byteSize()
        }
        return size
    }

    public func encode(writer: inout BinaryWriter) throws {
        guard entries.count <= Int(UInt16.max) else {
            throw WireFormatError.tooManyElements(entries.count)
        }
        try UInt16(entries.count).encode(writer: &writer)
        let sorted = entries.sorted { $0.key < $1.key }
        for (key, value) in sorted {
            try key.encode(writer: &writer)
            try value.encode(writer: &writer)
        }
    }

    public static func decode(reader: inout BinaryReader) throws -> OrderedMap<K, V> {
        let count = try UInt16.decode(reader: &reader)
        var entries: [(key: K, value: V)] = []
        entries.reserveCapacity(Int(count))
        for _ in 0..<count {
            let key = try K.decode(reader: &reader)
            let value = try V.decode(reader: &reader)
            entries.append((key: key, value: value))
        }
        return OrderedMap(entries)
    }
}

// MARK: - OrderedSet (BTreeSet equivalent)

// r[impl jetstream.wireformat.set]
// r[impl jetstream.wireformat.swift.set]
/// An ordered set that encodes elements in sorted order.
public struct OrderedSet<T: WireFormat & Comparable & Hashable>: Equatable {
    public var elements: [T]

    public init() {
        self.elements = []
    }

    public init(_ elements: [T]) {
        self.elements = elements.sorted()
    }
}

extension OrderedSet: WireFormat {
    public func byteSize() -> UInt32 {
        var size: UInt32 = 2
        for element in elements {
            size += element.byteSize()
        }
        return size
    }

    public func encode(writer: inout BinaryWriter) throws {
        guard elements.count <= Int(UInt16.max) else {
            throw WireFormatError.tooManyElements(elements.count)
        }
        try UInt16(elements.count).encode(writer: &writer)
        let sorted = elements.sorted()
        for element in sorted {
            try element.encode(writer: &writer)
        }
    }

    public static func decode(reader: inout BinaryReader) throws -> OrderedSet<T> {
        let count = try UInt16.decode(reader: &reader)
        var elements: [T] = []
        elements.reserveCapacity(Int(count))
        for _ in 0..<count {
            elements.append(try T.decode(reader: &reader))
        }
        return OrderedSet(elements)
    }
}

File: packages/jetstream_wireformat/src/collections.ts
/**
 * r[impl jetstream.wireformat.vec]
 * r[impl jetstream.wireformat.ts.array]
 * r[impl jetstream.wireformat.data]
 * r[impl jetstream.wireformat.ts.data]
 * r[impl jetstream.wireformat.map]
 * r[impl jetstream.wireformat.ts.map]
 * r[impl jetstream.wireformat.set]
 * r[impl jetstream.wireformat.ts.set]
 */

import { BinaryReader } from './binary-reader.js';
import { BinaryWriter } from './binary-writer.js';
import type { WireFormat } from './primitives.js';

const MAX_DATA_LENGTH = 32 * 1024 * 1024; // 32 MB

/**
 * Creates a Vec/Array codec with u16 count prefix.
 * Max 65,535 elements.
 */
export function vecCodec<T>(elementCodec: WireFormat<T>): WireFormat<T[]> {
  return {
    byteSize(value: T[]): number {
      let size = 2; // u16 count
      for (const elem of value) {
        size += elementCodec.byteSize(elem);
      }
      return size;
    },
    encode(value: T[], writer: BinaryWriter): void {
      if (value.length > 0xFFFF) {
        throw new Error(`too many elements in vector: ${value.length} exceeds u16 max (65535)`);
      }
      writer.writeU16(value.length);
      for (const elem of value) {
        elementCodec.encode(elem, writer);
      }
    },
    decode(reader: BinaryReader): T[] {
      const len = reader.readU16();
      const result: T[] = [];
      for (let i = 0; i < len; i++) {
        result.push(elementCodec.decode(reader));
      }
      return result;
    },
  };
}

/**
 * Data codec: u32 length prefix followed by raw bytes.
 * Max 32 MB (33,554,432 bytes).
 */
export const dataCodec: WireFormat<Uint8Array> = {
  byteSize(value: Uint8Array): number {
    return 4 + value.byteLength;
  },
  encode(value: Uint8Array, writer: BinaryWriter): void {
    writer.writeU32(value.byteLength);
    writer.writeBytes(value);
  },
  decode(reader: BinaryReader): Uint8Array {
    const len = reader.readU32();
    if (len > MAX_DATA_LENGTH) {
      throw new Error(`data length (${len} bytes) is too large, max is ${MAX_DATA_LENGTH}`);
    }
    return reader.readBytes(len);
  },
};

/**
 * Creates a Map codec with u16 count prefix.
 * Entries are encoded in sorted key order to match Rust BTreeMap.
 * Max 65,535 entries.
 */
export function mapCodec<K, V>(
  keyCodec: WireFormat<K>,
  valueCodec: WireFormat<V>,
  compareKeys: (a: K, b: K) => number,
): WireFormat<Map<K, V>> {
  return {
    byteSize(value: Map<K, V>): number {
      let size = 2; // u16 count
      for (const [k, v] of value) {
        size += keyCodec.byteSize(k) + valueCodec.byteSize(v);
      }
      return size;
    },
    encode(value: Map<K, V>, writer: BinaryWriter): void {
      if (value.size > 0xFFFF) {
        throw new Error(`map too large: ${value.size} exceeds u16 max (65535)`);
      }
      writer.writeU16(value.size);
      // Sort entries by key to match BTreeMap
      const entries = [...value.entries()].sort((a, b) => compareKeys(a[0], b[0]));
      for (const [k, v] of entries) {
        keyCodec.encode(k, writer);
        valueCodec.encode(v, writer);
      }
    },
    decode(reader: BinaryReader): Map<K, V> {
      const len = reader.readU16();
      const map = new Map<K, V>();
      for (let i = 0; i < len; i++) {
        const key = keyCodec.decode(reader);
        const value = valueCodec.decode(reader);
        map.set(key, value);
      }
      return map;
    },
  };
}

/**
 * Creates a Set codec with u16 count prefix.
 * Elements are encoded in sorted order to match Rust BTreeSet.
 * Max 65,535 elements.
 */
export function setCodec<T>(
  elementCodec: WireFormat<T>,
  compareElements: (a: T, b: T) => number,
): WireFormat<Set<T>> {
  return {
    byteSize(value: Set<T>): number {
      let size = 2; // u16 count
      for (const elem of value) {
        size += elementCodec.byteSize(elem);
      }
      return size;
    },
    encode(value: Set<T>, writer: BinaryWriter): void {
      if (value.size > 0xFFFF) {
        throw new Error(`set too large: ${value.size} exceeds u16 max (65535)`);
      }
      writer.writeU16(value.size);
      const sorted = [...value].sort(compareElements);
      for (const elem of sorted) {
        elementCodec.encode(elem, writer);
      }
    },
    decode(reader: BinaryReader): Set<T> {
      const len = reader.readU16();
      const set = new Set<T>();
      for (let i = 0; i < len; i++) {
        set.add(elementCodec.decode(reader));
      }
      return set;
    },
  };
}
